lib/Driver/ToolChain.cpp

```diff
@@ -150,6 +151,150 @@ types::ID ToolChain::lookupTypeForExtension(StringRef Ext) const {
   return types::lookupTypeForExtension(Ext);
 }
 
+// Return a _single_ TY_Swift InputAction, if one exists;
+// if 0 or >1 such inputs exist, return nullptr.
+static const InputAction*
+findSingleSwiftInput(const CompileJobAction *CJA) {
+  auto Inputs = CJA->getInputs();
+  const InputAction *IA = nullptr;
+  for (auto const *I : Inputs) {
+    if (auto const *S = dyn_cast<InputAction>(I)) {
+      if (S->getType() == types::TY_Swift) {
+        if (IA == nullptr) {
+          IA = S;
+        } else {
+          // Already found one, two is too many.
+          return nullptr;
+        }
+      }
+    }
+  }
+  return IA;
+}
+
+bool
+ToolChain::jobIsBatchable(const Job *A) const {
+  // FIXME: There might be a tighter criterion to use here?
+  auto const *CJActA = dyn_cast<const CompileJobAction>(&A->getSource());
+  if (!CJActA)
+    return false;
+  return findSingleSwiftInput(CJActA) != nullptr;
+}
+
+bool
+ToolChain::jobsAreBatchCombinable(const Job *A, const Job *B) const {
+
+  // Check that we have two CompileJobActions.
+  auto const *CJActA = dyn_cast<const CompileJobAction>(&A->getSource());
+  auto const *CJActB = dyn_cast<const CompileJobAction>(&B->getSource());
+  if (!CJActA || !CJActB)
+    return false;
+
+  // Check that we have two "single Swift input" jobs (possibly with other
+  // auxiliary inputs such as PCHs).
+  auto const *InActA = findSingleSwiftInput(CJActA);
+  auto const *InActB = findSingleSwiftInput(CJActB);
+  if (!InActA || !InActB)
+    return false;
+
+  // Check Jobs have same executable.
+  if (strcmp(A->getExecutable(), B->getExecutable()) != 0)
+    return false;
+
+  // Check Jobs are making the same kind of output.
+  if (A->getOutput().getPrimaryOutputType() !=
+      B->getOutput().getPrimaryOutputType())
+    return false;
+
+  // Check Jobs have same environment.
+  auto AEnv = A->getExtraEnvironment();
+  auto BEnv = B->getExtraEnvironment();
+  if (AEnv.size() != BEnv.size())
+    return false;
+  for (size_t i = 0; i < AEnv.size(); ++i) {
+    if (strcmp(AEnv[i].first, BEnv[i].first) != 0)
+      return false;
+    if (strcmp(AEnv[i].second, BEnv[i].second) != 0)
+      return false;
+  }
+
+  return true;
+}
+
+std::unique_ptr<Job>
+ToolChain::constructBatchJob(ArrayRef<const Job *> jobs,
+                             Compilation &C) const
+{
+  // Here we construct an aggregate of a set of jobs; a precondition of
+  // this is that the jobs are all pairwise combinable.
+#ifndef NDEBUG
+  for (auto *A : jobs) {
+    for (auto *B : jobs) {
+      assert(jobsAreBatchCombinable(A, B));
+    }
+  }
+#endif
+  // As much as possible, we treat the construction of the batch job the
+  // same as we did the constituent jobs, building an aggregate
+  // CompileJobAction and calling back into constructInvocation as before,
+  // with a CompileJobAction and JobContext that differ only slightly.
+
+  if (jobs.size() == 0)
+    return nullptr;
```

Comment:
If it's 1, worth early-exiting with the single Job?

Reply:
Hm, I don't think so; but I do think we should guard against _getting_ to this path with any number of jobs < 2. I'll do a little refactoring in the caller and put an assert in here.

Reply:
Eh actually this just moves ugliness in the logic elsewhere, I'm going to leave it as-is.